flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -0,0 +1,410 @@
+      rightCache.put(timeForRightRow, rightRowList)
+      if (leftTimerState.value == 0) {
+        // Register a timer on the LEFT stream to remove rows.
+        registerCleanUpTimer(ctx, timeForRightRow, leftTimerState, leftRow = false)
+      }
+    }
+    // We'd like to produce as many results as possible.
+    if (leftExpirationTime < leftQualifiedUpperBound) {
+      leftExpirationTime = calExpirationTime(rightOperatorTime, leftRelativeSize)
+      // Join the rightRow with rows from the left cache.
+      val leftIterator = leftCache.iterator()
+      while (leftIterator.hasNext) {
+        val leftEntry = leftIterator.next
+        val leftTime = leftEntry.getKey
+        if (leftTime >= leftQualifiedLowerBound && leftTime <= leftQualifiedUpperBound) {
+          val leftRows = leftEntry.getValue
+          var i = 0
+          while (i < leftRows.size) {
+            joinFunction.join(leftRows.get(i), rightRow, cRowWrapper)
+            i += 1
+          }
+        }
+        if (leftTime <= leftExpirationTime) {
+          // eager remove
+          leftIterator.remove()
+        } // We could do the short-cutting optimization here once we get a state with ordered keys.
+      }
+    }
+  }
+
+  /**
+    * Called when a registered timer is fired.
+    * Remove rows whose timestamps are earlier than the expiration time,
+    * and register a new timer for the remaining rows.
+    *
+    * @param timestamp the timestamp of the timer
+    * @param ctx       the context to register timer or get current time
+    * @param out       the collector for returning result values
+    */
+  override def onTimer(
+      timestamp: Long,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+      out: Collector[CRow]): Unit = {
+    updateOperatorTime(ctx)
+    // In the future, we should separate the left and right watermarks. Otherwise, the
+    // registered timer of the faster stream will be delayed, even if the watermarks have
+    // already been emitted by the source.
+    if (leftTimerState.value == timestamp) {
+      rightExpirationTime = calExpirationTime(leftOperatorTime, rightRelativeSize)
+      removeExpiredRows(
+        rightExpirationTime,
+        rightCache,
+        leftTimerState,
+        ctx,
+        removeLeft = false
+      )
+    }
+
+    if (rightTimerState.value == timestamp) {
+      leftExpirationTime = calExpirationTime(rightOperatorTime, leftRelativeSize)
+      removeExpiredRows(
+        leftExpirationTime,
+        leftCache,
+        rightTimerState,
+        ctx,
+        removeLeft = true
+      )
+    }
+  }
+
+  /**
+    * Calculate the expiration time with the given operator time and relative window size.
+    *
+    * @param operatorTime the operator time
+    * @param relativeSize the relative window size
+    * @return the expiration time for cached rows
+    */
+  private def calExpirationTime(operatorTime: Long, relativeSize: Long): Long = {
+    if (operatorTime < Long.MaxValue) {
```

Comment:
Since `relativeSize` may be negative, the expression `operatorTime - relativeSize - allowedLateness - 1` will overflow when `operatorTime = Long.MaxValue` and `relativeSize < 0`. When encountered this, there will be endless timers with negative expiration times (which should have been `Long.MaxValue`) being recursively registered and the job will never stop.

Reply:
OK, but the current implementation does not prevent overflows.
Shouldn't the method rather be:

```
val expirationTime = operatorTime - relativeSize - allowedLateness - 1
if (expirationTime >= 0) {
  expirationTime
} else {
  Long.MaxValue
}
```

Reply:
Thanks for pointing this out, @fhueske. Actually the `expirationTime` can be negative when the job has just been started (e.g., when operatorTime = 10 and relativeSize = 20). I'd rather take the Long.MaxValue for `operatorTime` as a special case and keep the original logic. What do you think?

Reply:
But what does the current check prevent then? We might still run in an overflow if `relativeSize` is negative. Moreover, the case of `operatorTime == Long.MaxValue` may only happen at the end of the job when no more records will be received. In this case the overflow would have no consequence, right? Doesn't it make more sense to prevent the registration of negative timers?

Reply:
There are two cleanup-related times: (1) the expected cleanup time, which is decided by the rowtime, used for registering the timer and triggered by the operator time; and (2) the real expiration time, which is calculated by the operator time (the code snippet we are discussing). In most cases, (1) the expected cleanup time will not overflow (i.e. the real rowtime will never approach `Long.MaxValue`) and (2) the expiration time calculated in the `onTimer()` method should be greater than or equal to some timestamps of the cached records, otherwise there will be no records to be removed.

There is only one situation we should pay attention to: the expected cleanup time > 0, while the real expiration time < 0 (caused by overflow). In that case, no records will be removed, thus we register a new timer with a positive timestamp. This timer will be triggered by the `Long.MaxValue` operator time, but again no records will be removed and a new timer will be registered. It's a dead loop.

Reply:
I'll change the condition here to be more precise.

Reply:
If we still want to check the overflow, the condition need to be more precise.
```
val expirationTime = operatorTime - relativeSize - allowedLateness - 1
if (expirationTime < 0 && relativeSize + allowedLateness + 1 < 0) {
    Long.MaxValue
} else {
    expirationTime
}
```
If we just check the end of the stream and ignore the other overflow circumstances, I think the original code also works.
```
if (operatorTime < Long.MaxValue) {
    operatorTime - relativeSize - allowedLateness - 1
} else {
    Long.MaxValue
}
```
Which one do you prefer?

Reply:
Thanks for the explanation. Let's keep the current implementation.